mautic/mautic
# Lead filter query optimization

This is a duplicate of https://github.com/mautic/mautic/pull/9944 with 5.x merged to resolve the conflicts

> Q	A
> Branch?	"features"
> Bug fix?	no
> New feature?	no
> Deprecations?	yes
> BC breaks?	no
> Automated tests included?	yes
> Related user documentation PR URL	
> Related developer documentation PR URL	
> Issue(s) addressed	
> #### Description:
> This PR optimizes DB queries that are executed on the contact list page when searching by the `segment:{segment_alias}` keyword with segments having millions of contacts. The optimization includes:
> 
> 1. Replacing the join of `mautic_lead_lists` table by a separate query.
> 2. Getting rid of expensive `DISTINCT` and `GROUP BY` clauses by rewriting with a subquery.
> 3. Adding an index hint.
> 4. Removal of duplicate code in `\Mautic\LeadBundle\Segment\Query\QueryBuilder`.
> 
> Original query:
> 
> ```sql
> SELECT l.* FROM mautic_leads l
>     LEFT JOIN mautic_users u ON u.id = l.owner_id
>     INNER JOIN mautic_lead_lists_leads list_lead ON l.id = list_lead.lead_id
>     INNER JOIN mautic_lead_lists list ON list_lead.leadlist_id = list.id
> WHERE ((list_lead.manually_removed = 0) AND (list.alias = "segment_alias"))
>     AND (l.date_identified IS NOT NULL)
> GROUP BY l.id
> ORDER BY l.last_active DESC
> LIMIT 30
> ```
> 
> Optimized query:
> 
> ```sql
> SELECT id FROM mautic_lead_lists WHERE alias = "segment_alias";
> SELECT l.* FROM mautic_leads l USE INDEX FOR JOIN (`mautic_lead_date_added`)
>     LEFT JOIN mautic_users u ON u.id = l.owner_id
> WHERE (EXISTS(SELECT 1
>     FROM mautic_lead_lists_leads lla
>     WHERE (l.id = lla.lead_id)
>         AND (lla.manually_removed = 0)
>         AND (lla.leadlist_id IN (/* IDs returned by the first query */ )))) 
>     AND (l.date_identified IS NOT NULL)
> ORDER BY l.last_active DESC
> LIMIT 30;
> ```
> 
> #### Steps to test this PR:
> 1. Create a segments that contains about 4M contacts.
> 2. Go to the segment list page.
> 3. Click the `View 4000000 Contacts` link.
> 4. The contact page (with the filter `segment:{segment_alias}`) should be loaded in about 15 seconds (Before it took about 2 minutes).
> 
> #### List deprecations
> 1. Method `\Mautic\LeadBundle\Segment\Query\QueryBuilder::getJoinCondition()`
> 2. Method `\Mautic\LeadBundle\Segment\Query\QueryBuilder::replaceJoinCondition()`
